File: python/design-patterns/dependency-injection.py
from abc import ABC, abstractmethod

class Service(ABC):
    @abstractmethod
    def execute(self):
        """Abstract method representing the execution of a service."""
        pass

class Client:
    def __init__(self, service):
        """
        Constructor for the Client class.

        Args:
            service (Service): The service to be used by the client.
        """
        self.service = service

    def do_something(self):
        """Performs an action using the provided service."""
        self.service.execute()

class ConcreteService(Service):
    def execute(self):
        """Concrete implementation of the service execution."""
        print("Executing ConcreteService")

def main():
    # Create a concrete service
    service = ConcreteService()
    # Create a client with the concrete service
    client = Client(service)
    # Perform an action using the service
    client.do_something()

if __name__ == "__main__":
    main()

File: cpp/design-patterns/dependency-injection/dependency-injection.cpp
#include <iostream>
#include <memory>

/**
 * @brief The Service interface declares a method for executing a service.
 */
class Service {
public:
    virtual void execute() const = 0;

    /**
     * @brief Virtual destructor to allow proper destruction of derived classes
     */
    virtual ~Service() = default;
};

/**
 * @brief The Client class depends on the Service interface, not concrete
 * implementations.
 */
class Client {
private:
    std::unique_ptr<Service> service; /**< Pointer to the Service object */

public:
    /**
     * @brief Construct a new Client object with the provided Service.
     * 
     * @param s Unique pointer to the Service object
     */
    Client(std::unique_ptr<Service> s) : service(std::move(s)) {}

    /**
     * @brief Performs an action using the provided Service.
     */
    void doSomething() const {
        service->execute();
    }
};

/**
 * @brief The ConcreteService class provides a concrete implementation of the
 * Service interface.
 */
class ConcreteService : public Service {
public:
    /**
     * @brief Executes the concrete service.
     */
    void execute() const override {
        std::cout << "Executing ConcreteService\n";
    }
};

/**
 * @brief The main function demonstrates the use of the Dependency Inversion
 * Principle (DIP) in SOLID.
 * 
 * @return int Program exit status
 */
int main() {
    std::unique_ptr<Service> service = std::make_unique<ConcreteService>();
    std::unique_ptr<Client> client = std::make_unique<Client>(std::move(service));
    client->doSomething();
    return 0;
}
